flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TriggerManager.java

```diff
@@ -0,0 +1,295 @@
+    this.lockFactory = lockFactory;
+    this.taskNames = taskNames;
+    this.evaluators = evaluators;
+    this.minFireDelayMs = minFireDelayMs;
+    this.lockCheckDelayMs = lockCheckDelayMs;
+    this.clearLocks = clearLocks;
+  }
+
+  @Override
+  public void open(Configuration parameters) throws Exception {
+    this.rateLimiterTriggeredCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(MetricConstants.GROUP_KEY, MetricConstants.GROUP_VALUE_DEFAULT)
+            .counter(MetricConstants.RATE_LIMITER_TRIGGERED);
+    this.concurrentRunTriggeredCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(MetricConstants.GROUP_KEY, MetricConstants.GROUP_VALUE_DEFAULT)
+            .counter(MetricConstants.CONCURRENT_RUN_TRIGGERED);
+    this.nothingToTriggerCounter =
+        getRuntimeContext()
+            .getMetricGroup()
+            .addGroup(MetricConstants.GROUP_KEY, MetricConstants.GROUP_VALUE_DEFAULT)
+            .counter(MetricConstants.NOTHING_TO_TRIGGER);
+    this.triggerCounters =
+        taskNames.stream()
+            .map(
+                name ->
+                    getRuntimeContext()
+                        .getMetricGroup()
+                        .addGroup(MetricConstants.GROUP_KEY, name)
+                        .counter(MetricConstants.TRIGGERED))
+            .collect(Collectors.toList());
+
+    this.nextEvaluationTime =
+        getRuntimeContext()
+            .getState(new ValueStateDescriptor<>("triggerManagerNextTriggerTime", Types.LONG));
+    this.accumulatedChanges =
+        getRuntimeContext()
+            .getListState(
+                new ListStateDescriptor<>(
+                    "triggerManagerAccumulatedChange", TypeInformation.of(TableChange.class)));
+    this.lastTriggerTimes =
+        getRuntimeContext()
+            .getListState(new ListStateDescriptor<>("triggerManagerLastTriggerTime", Types.LONG));
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    // Do nothing
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) {
+    LOG.info("Initializing state restored: {}, clearLocks: {}", context.isRestored(), clearLocks);
+    this.lock = lockFactory.createLock();
+    this.recoveryLock = lockFactory.createRecoveryLock();
+    if (context.isRestored()) {
+      isCleanUp = true;
+    } else if (clearLocks) {
+      // Remove old lock if we are not restoring the job
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public void processElement(TableChange change, Context ctx, Collector<Trigger> out)
+      throws Exception {
+    long current = ctx.timerService().currentProcessingTime();
+    Long nextTime = nextEvaluationTime.value();
+    init(out, ctx.timerService(), current);
+
+    // Add the new changes to the already accumulated ones
+    List<TableChange> accumulated = Lists.newArrayList(accumulatedChanges.get());
```

Comment:
This is the `TableChange` per scheduled Maintenance Task since the last Trigger of the given task.

Reply:
still not quite following. I thought TriggerManager accumulates `TableChange` and decide if it is time to actually trigger/schedule the maintenance task.

Reply:
We're planning to have multiple Maintenance Tasks in the PostCommitTopology. But we are planning to have a single `TriggerManager`, so it is easier to prevent concurrent runs. This single `TriggerManager` needs to handle the scheduling of every Maintenance Task. Every Maintenance Task could have a different scheduling requirement, like DataFileCompaction on ever 10 commits, ExpireSnapshots  on every hour, and DeleteOrphanFiles on every day. For this, we need to keep track the accumulated `TableChange`s for every Maintenance Task.

OTOH some users might want to schedule Maintenance Tasks in a group, like every 10 commits do a DataFileCompaction, then do a ManifestFileCompaction, then do an ExpireSnapshots. If all of them has independent scheduling,  then the order is not guaranteed, and also there fight for the lock.

That's why I opted for the single TriggerManager approach.